File: day05/01.js
const fs = require('fs')

function match(s1, s2) {
  return s1.toLowerCase() == s2.toLowerCase() && s1 !== s2
}

function produce(str, index = 0) {
  while (true) {
    if (str.length - 1 === index + 1) {
      return str
    }

    if (match(str.charAt(index), str.charAt(index + 1))) {
      const front = str.slice(0, index)
      str = front + str.slice(index + 2)

      index = Math.max(0, index - 1)
    } else {
      index += 1
    }
  }
}

// const data = fs.readFileSync('sample.txt', 'utf8')
const data = fs.readFileSync('input.txt', 'utf8')

const ans = produce(data).length
console.log(ans)

File: day05/01.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var minV = 999999999

func match(s1, s2 string) bool {
	return strings.ToLower(s1) == strings.ToLower(s2) && s1 != s2
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func produce(str string, index int) string {
	if len(str)-1 == index+1 {
		return str
	}

	if match(string(str[index]), string(str[index+1])) {
		front := str[:index]
		newStr := front + str[index+2:]

		return produce(newStr, max(0, index-1))
	}
	return produce(str, index+1)
}

func producer(s string) int {
	return len(produce(s, 0))
}

func main() {
	file, _ := os.Open("input.txt")
	// file, _ := os.Open("sample.txt")

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var str string

	for scanner.Scan() {
		str = scanner.Text()
	}

	fmt.Println(producer(str))

}
